ndarray/src/main/java/org/tensorflow/ndarray/impl/sparse/AbstractSparseNdArray.java

```diff
@@ -0,0 +1,481 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.ndarray.impl.sparse;
+
+import org.tensorflow.ndarray.IllegalRankException;
+import org.tensorflow.ndarray.LongNdArray;
+import org.tensorflow.ndarray.NdArray;
+import org.tensorflow.ndarray.NdArraySequence;
+import org.tensorflow.ndarray.NdArrays;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.ndarray.SparseNdArray;
+import org.tensorflow.ndarray.StdArrays;
+import org.tensorflow.ndarray.impl.AbstractNdArray;
+import org.tensorflow.ndarray.impl.dense.AbstractDenseNdArray;
+import org.tensorflow.ndarray.impl.dimension.Dimension;
+import org.tensorflow.ndarray.impl.dimension.DimensionalSpace;
+import org.tensorflow.ndarray.impl.dimension.RelativeDimensionalSpace;
+import org.tensorflow.ndarray.impl.sequence.SingleElementSequence;
+import org.tensorflow.ndarray.impl.sequence.SlicingElementSequence;
+import org.tensorflow.ndarray.index.Index;
+
+import java.nio.ReadOnlyBufferException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.LongStream;
+
+/**
+ * Abstract base class for sparse array.
+ *
+ * <p>A sparse array as two separate dense arrays: indices, values, and a shape that represents the
+ * dense shape.
+ *
+ * <p><em>NOTE:</em> all Sparse Arrays are readonly for the {@link #set(NdArray, long...)} and
+ * {@link #setObject(Object, long...)} methods
+ *
+ * <pre>{@code
+ * FloatSparseNdArray st = new FloatSparseNdArray(
+ *      StdArrays.of(new long[][] {{0, 0}, {1, 2}},
+ *      StdArrays.of(new float[] {1f, 2f},
```

Comment:
Whenever possible in your example, I would suggest to use `NdArrays` instead of passing by `StdArrays`, which is slower. Something like:
```
new FloatSparseNdArray(
    StdArrays.of(new long[][] {{0, 0}, {1, 2}}), <----- missing closing parentheses by the way...
    NdArrays.ofFloats(1f, 2f),
    Shape.of(3, 4)
);

Reply:
Actually `NdArrays.ofFloats(...)` is wrong, it should be `NdArrays.vectorOf(...)`,  for the values. I have changed the values argument to `NdArrays.vectorOf(...)`, and fixed closing parenthesis. Also, changed 
`setValues(StdArrays.ndCopyOf(valuesArray))` to `setValues(NdArrays.vectorOf(valuesArray))` in the `write(ByteDataBuffer)`  methods.